addon/array/fragment.js

```diff
@@ -81,15 +31,45 @@ const FragmentArray = StatefulArray.extend({
 
   options: null,
 
-  /**
-    @method _normalizeData
-    @private
-    @param {Object} data
-  */
-  _normalizeData(data) {
-    let content = this.content;
+  objectAt(index) {
+    const recordData = this._super(index);
+    if (recordData === undefined) {
+      return;
+    }
+    const internalModel = this.store._internalModelForResource(recordData.identifier);
+    if (internalModel === undefined) {
+      return;
+    }
+    return internalModel.getRecord();
+  },
+
+  _normalizeData(data, index) {
+    assert(
+      `You can only add '${this.modelName}' fragments or object literals to this property`,
+      typeOf(data) === 'object' || isInstanceOfType(this.store.modelFor(this.modelName), data)
+    );
+
+    if (isFragment(data)) {
+      const recordData = recordDataFor(data);
+      setFragmentOwner(data, this.recordData, this.key);
+      return recordData;
+    }
+    const existing = this.objectAt(index);
```

Comment:
This is the spot where sending index(0) in my suggested case will cause duplication. We don't want to actually modify the first object in the array, hence the `start + i` addition.

Reply:
Good find!